Liste des livres, sélection d'un livre et détail

geekarist opened the pull request:
- `ListBookActivity` : contient une `RecyclerView` pour afficher les livres récupérés par Retrofit, et une autre vue dans un `RelativeLayout` qui permet de recharger la liste au cas où il y a un problème de connexion. Design "inspiré" de Google Play, les items sont affichés dans une `CardView`
- `ListAdapter` : utilise Glide pour afficher l'image de chaque livre, avec un placeholder très astucieux
- `BookDetailActivity` : n'affiche que le titre pour l'instant, mais c'est là que je prévois de mettre le bouton Acheter et l'image de couverture en grand

geekarist committed:
Create application with a `RecyclerView`

geekarist committed:
Display `Book`s in adapter

geekarist committed:
Use `CardView`

geekarist committed:
Load book cover into image view

geekarist committed:
Fix book covers not downloading

Add internet permission

geekarist committed:
Fetch list of books from web service

geekarist committed:
Display price

geekarist committed:
Show reload button on error when fetching books

geekarist committed:
Add click listener on list items

geekarist committed:
Send book as parcelable extra to detail activity

geekarist committed:
Improve layout of book detail

geekarist committed:
Display book cover in detail

geekarist committed:
Improve detail layout

geekarist committed:
Display price in buy button

geekarist committed:
Display cart when clicking on back button

geekarist committed:
Fix NPE when starting DisplayCartActivity

geekarist committed:
Display v7 toolbar in cart

geekarist committed:
Add an Up action to CartActivity

geekarist committed:
Add RelativeLayout for cart total

geekarist committed:
Calculate and display cart total

geekarist committed:
Add cart checkout button

geekarist committed:
Save and restore cart

Not working for now.

- Save the cart in `onSaveInstanceState`
- Restore it in `onCreate`

geekarist committed:
Fix wrong traduction

blacroix commented on app/build.gradle:
```diff
@@ -0,0 +1,32 @@
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
```
Cette ligne là est superflue dans le cas où tu n'importes rien depuis le dossier lib

blacroix commented on app/src/main/java/com/github/geekarist/hp2/ListBooksActivity.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private void fetchBooks() {
+        Retrofit retrofit = new Retrofit.Builder().baseUrl("http://henri-potier.xebia.fr")
```
Tu pers en performance à initialiser retrofit et le service avant chaque appel réseau. Tu peux le faire dans le onCreate() ou sinon dans l'Application (à créer).

blacroix commented on app/src/main/java/com/github/geekarist/hp2/ListBooksActivity.java:
```diff
@@ -0,0 +1,73 @@
+        bookService.listBooks().enqueue(new Callback<List<Book>>() {
+            @Override
+            public void onResponse(Response<List<Book>> response, Retrofit retrofit) {
+                mBookListAdapter.setBooks(response.body());
```
L'appel réseau se fait sur un autre thread, ici, il y a une fuite mémoire car tu empêches l'activité de se détruire. Je te conseille d'utiliser une classe interne statique et une WeakReference sur l'activité. Une fois l'appel réseau terminé regarde si l'activité n'est pas nul et si elle n'est pas en train de se terminer (isFinishing()) dans ce cas la mise à jour de la vue est inutile.

blacroix commented on app/src/main/java/com/github/geekarist/hp2/ListBooksAdapter.java:
```diff
@@ -0,0 +1,93 @@
+                .load(book.cover)
+                .placeholder(R.drawable.book_cover_placeholder)
+                .into(holder.mImageView);
+        holder.itemView.setOnClickListener(new View.OnClickListener() {
```
Pour des raisons de performance l'ajout de l'événement est clique peut être fait dans la méthode onCreateViewHolder() et non pas à chaque onBindViewHolder()

geekarist replied:
Dans ce cas, comment lui transmettre le `Book` sélectionné ? La position de l'item est fournie dans `onBindViewHolder`.

blacroix replied:
Dans la classe ViewHolder rien ne t'empêche d'ajouter une variable d'instance Book que tu changes à chaque onBindViewHolder()

geekarist replied:
Ca ne marche pas comme ça, au moment du clic la valeur du Book n'est pas la bonne.
Ca marche mieux si c'est fait dans le ViewHolder.

blacroix replied:
Faudra me montrer car en gardant une variable d'instance dans le ViewHolder il n'y a pas de raison que ça ne marche pas si tu changes bien cette variable à chaque bind.

blacroix commented on app/src/main/res/layout/activity_list_books.xml:
```diff
@@ -0,0 +1,33 @@
+        android:layout_height="match_parent"
+        android:scrollbars="vertical" />
+
+    <RelativeLayout
```
Le LinearLayout est à préférer quand c'est possible, car le RelativeLayout est plus gourmand sur les anciennes versions d'Android.

blacroix commented on app/src/main/res/layout/book_item.xml:
```diff
@@ -0,0 +1,46 @@
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <android.support.v7.widget.CardView
```
La CardView peut être le root element de ce layout non ? En général limiter la profondeur des layouts est une bénéfique pour la performance et la maintenabilité.

geekarist committed:
Add TODO to fix cart restoration

Commit:
Fix cart save/restore

- Save in `onStop` by serializing the list of books using `Gson`
- Restore in `onCreate` by deserializing

Commit:
Remove `fileTree` dependencies

Commit:
Initialize retrofit in `onCreate`

Commit:
Fix memory leak when fetching books

- Use a static inner `FetchBookCallback` class
- The callback has a weak reference to the activity
- It checks for a null or finishing activity before updating the view

Commit:
Use `LinearLayout` instead of `RelativeLayout`

For performance reasons.

Commit:
Set `CardView` as top level element in book item layout

Commit:
Move `setOnClickListener` to `ViewHolder`

For performance.

Commit:
Get context from `ViewHolder.itemView`

Commit:
Increase text size

Commit:
Fix cart layout for total

Commit:
Display cart price before and after discount

Commit:
Implement best offer calculation (fake discounts)
